pkg/api/daemon.go

```diff
@@ -65,26 +65,44 @@ func Run(context *clusterd.Context, c *Config) error {
 }
 
 func WatchMonConfig(context *clusterd.Context, c *Config) {
-	opts := metav1.ListOptions{
-		FieldSelector: fields.OneTermEqualSelector(api.ObjectNameField, monop.EndpointConfigMapName).String(),
-	}
-	w, err := context.Clientset.Core().ConfigMaps(c.namespace).Watch(opts)
-	if err != nil {
-		logger.Errorf("API server init error: %+v", err)
-	}
-	defer w.Stop()
-
+	// infinite loop to keep watching forever, this function should be called in a goroutine
 	for {
-		e := <-w.ResultChan()
-		if e.Type == watch.Modified {
-			// "unmarshal" object into configmap and set new endpoints
-			monEndpoints := e.Object.(*v1.ConfigMap)
-			c.clusterInfo.Monitors = mon.ParseMonEndpoints(monEndpoints.Data[monop.EndpointDataKey])
-
-			// write the latest config to the config dir
-			if err := mon.GenerateAdminConnectionConfig(context, c.clusterInfo); err != nil {
-				logger.Errorf("failed to write connection config. %+v", err)
-				return
+		// watch for changes only to the monitor endpoints config map
```

Comment:
why not have a conditional to check whether the watch object is invalid and then recreate one? Something like this

```
e, ok := <-w.ResultChan()
if !ok {
    if w != nil {
        w.Stop()
    }
    // create the watch again
    ...
    continue // start loop again
}
```

Reply:
with that approach, we'd either need to refactor the creation of the watcher into a new function, so it can be called at the beginning of this function and when the result channel is closed, or just duplicate the watcher creation in the 2 places.  I don't see either as a strong improvement over the current implementation, do you?

Reply:
i thought it would make it a bit more readable. You wouldnt have nested for loops and no defer function. But you would have to extract the creation of the watch object into a separate function to not duplicate it like you said.

Im ok with this.

Reply:
but it wouldnt be too much since you would only need to call the creation of the watch object, which is:
```w, err := context.Clientset.Core().ConfigMaps(c.namespace).Watch(opts)```

Reply:
you have to handle the potential error from that watcher creation in 2 places.  you'd also still need the `defer` function in case you are returning from the entire function if `mon.GenerateAdminConnectionConfig` fails.  IMHO, i don't think this change is worth the perceived readability improvement.

Reply:
Yes. You still would need something like this at the beginning:
`defer w.Stop()` 
OK LGTM